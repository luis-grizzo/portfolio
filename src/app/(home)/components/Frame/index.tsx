'use client'

import Image from 'next/image'
import { motion } from 'framer-motion'

import { useColor } from '@/hooks/useColor'

import { verticalHidden, verticalVisible } from '@/utils/animations'

export const Frame = () => {
  const { color } = useColor()

  return (
    <motion.div
      initial={verticalHidden}
      animate={verticalVisible}
      exit={verticalHidden}
      className={`flex items-end justify-center relative w-full h-72 xsm:h-96 lg:h-full rounded-3xl overflow-hidden before:content-[''] before:bg-${color}-600 before:h-[65%] before:w-full before:rounded-3xl`}
    >
      <motion.div
        initial={{ y: '100%', opacity: 0 }}
        animate={{ y: '2%', opacity: 1 }}
        exit={{ y: '100%', opacity: 0 }}
        transition={{ type: 'spring', bounce: 0.2 }}
        className="absolute h-full w-max"
      >
        <Image
          priority
          width={2943}
          height={2022}
          className="h-full w-max"
          src="/portrait.png"
          alt="LuÃ­s Grizzo's portrait"
        />
      </motion.div>
    </motion.div>
  )
}
