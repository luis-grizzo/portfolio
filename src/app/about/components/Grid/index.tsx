'use client'

import { cloneElement, useState, useEffect, useRef } from 'react'
import { AnimatePresence, motion } from 'framer-motion'

import { useColor } from '@/hooks/useColor'

import { verticalHidden, verticalVisible } from '@/utils/animations'
import { techs } from '@/utils/techs'

export const Grid = (): React.ReactElement => {
  const gridRef = useRef<HTMLDivElement>(null)

  const { color } = useColor()

  const [isFullLeftScroll, setIsFullLeftScroll] = useState(false)
  const [isFullRightScroll, setIsFullRightScroll] = useState(false)

  const handleScroll = (): void => {
    const offsetWidth = gridRef.current?.offsetWidth ?? 0
    const scrollLeft = gridRef.current?.scrollLeft ?? 0
    const scrollWidth = gridRef.current?.scrollWidth ?? 0

    scrollLeft === 0 ? setIsFullLeftScroll(true) : setIsFullLeftScroll(false)

    offsetWidth + scrollLeft >= scrollWidth
      ? setIsFullRightScroll(true)
      : setIsFullRightScroll(false)
  }

  useEffect(() => {
    handleScroll()

    gridRef.current?.addEventListener('scroll', handleScroll)

    return () => gridRef.current?.removeEventListener('scroll', handleScroll)
  }, [])

  return (
    <div className="flex relative h-full w-full overflow-hidden">
      <div
        ref={gridRef}
        className="grid grid-cols-[repeat(12,_auto)] lg:grid-cols-[1fr_1fr_1fr] auto-rows-[1fr] gap-6 w-full h-full overflow-y-hidden"
      >
        {techs.map((tech, index) => (
          <motion.div
            key={tech.name}
            initial={verticalHidden}
            animate={verticalVisible}
            exit={verticalHidden}
            transition={{ delay: index * 0.1 }}
            className={`flex items-center justify-center w-40 aspect-square lg:aspect-auto lg:w-full rounded-[1.25rem] text-contrast_color_lightTheme dark:text-contrast_color_darkTheme bg-shape_color_lightTheme dark:bg-shape_color_darkTheme hover:bg-${color}-600 transition-colors`}
          >
            {cloneElement(tech.icon, {
              size: 60,
              className: 'gi__icon'
            })}
          </motion.div>
        ))}
      </div>

      <AnimatePresence>
        {!isFullLeftScroll && (
          <motion.div
            key="left"
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            exit={{ opacity: 0 }}
            className="absolute top-0 left-0 h-full w-1/6 bg-gradient-to-r from-background_color_lightTheme dark:from-background_color_darkTheme"
          />
        )}

        {!isFullRightScroll && (
          <motion.div
            key="right"
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            exit={{ opacity: 0 }}
            className="absolute top-0 right-0 h-full w-1/6 bg-gradient-to-l from-background_color_lightTheme dark:from-background_color_darkTheme"
          />
        )}
      </AnimatePresence>
    </div>
  )
}
