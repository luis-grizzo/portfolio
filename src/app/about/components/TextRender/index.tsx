'use client'

import { useState, useEffect, useRef } from 'react'
import { AnimatePresence, motion } from 'framer-motion'

import { horizontalHidden, horizontalVisible } from '@/utils/animations'

interface TextRenderProps {
  texts: TextProps[]
}

interface TextProps {
  id: number
  paragraph: React.ReactElement
}

export const TextRender = ({ texts }: TextRenderProps) => {
  const textContainerRef = useRef<HTMLDivElement>(null)

  const [isFullScrolled, setIsFullScrolled] = useState(false)
  const [isTopScroll, setIsTopScroll] = useState(false)

  const handleScroll = (): void => {
    const offsetHeight = textContainerRef.current?.offsetHeight ?? 0
    const scrollTop = textContainerRef.current?.scrollTop ?? 0
    const scrollHeight = textContainerRef.current?.scrollHeight ?? 0

    scrollTop === 0 ? setIsTopScroll(true) : setIsTopScroll(false)

    offsetHeight + scrollTop >= scrollHeight
      ? setIsFullScrolled(true)
      : setIsFullScrolled(false)
  }

  useEffect(() => {
    handleScroll()

    textContainerRef.current?.addEventListener('scroll', handleScroll)

    return () =>
      textContainerRef.current?.removeEventListener('scroll', handleScroll)
  }, [])

  return (
    <div className="flex relative h-[40%] overflow-hidden">
      <div
        ref={textContainerRef}
        className="flex flex-col items-start gap-6 overflow-auto overflow-x-hidden h-full w-full pr-2 text-xl text-text_color_lightTheme dark:text-text_color_darkTheme custom-scroll"
      >
        {texts.map((text, index) => (
          <motion.div
            key={text.id}
            initial={horizontalHidden}
            animate={horizontalVisible}
            exit={horizontalHidden}
            transition={{ delay: (index + 1) * 0.1 }}
          >
            {text.paragraph}
          </motion.div>
        ))}
      </div>

      <AnimatePresence mode="wait">
        {!isTopScroll && (
          <motion.div
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            exit={{ opacity: 0 }}
            className="absolute top-0 left-0 h-1/4 w-[calc(100%_-_0.5rem)] bg-gradient-to-b from-background_color_lightTheme to-transparent dark:from-background_color_darkTheme"
          />
        )}
      </AnimatePresence>

      <AnimatePresence mode="wait">
        {!isFullScrolled && (
          <motion.div
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            exit={{ opacity: 0 }}
            className="absolute bottom-0 left-0 h-1/4 w-[calc(100%_-_0.5rem)] bg-gradient-to-t from-background_color_lightTheme to-transparent dark:from-background_color_darkTheme"
          />
        )}
      </AnimatePresence>
    </div>
  )
}
