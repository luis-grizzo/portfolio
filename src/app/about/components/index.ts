export * from './Grid'
export * from './TextRender'
