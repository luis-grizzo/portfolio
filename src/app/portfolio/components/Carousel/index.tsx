'use client'

import { cloneElement, useState, useEffect } from 'react'
import Image from 'next/image'
import { AnimatePresence, motion } from 'framer-motion'
import { FaCss3 } from 'react-icons/fa6'
import { BiLogoTypescript } from 'react-icons/bi'
import { DiJavascript1 } from 'react-icons/di'
import { MdStar, MdCode } from 'react-icons/md'

import { useColor } from '@/hooks/useColor'
import { useResize } from '@/hooks/useResize'

import { verticalHidden, verticalVisible } from '@/utils/animations'

import { FetchReposProps } from '@/types'

interface CarouselProps {
  items: FetchReposProps[]
}

export const Carousel = ({ items }: CarouselProps) => {
  const { color } = useColor()
  const { isDesktop } = useResize()

  const [activeItem, setActiveItem] = useState(0)

  const handleLanguageIcon = (language: string) => {
    const selectedLanguage = {
      css: <FaCss3 />,
      typescript: <BiLogoTypescript />,
      javascript: <DiJavascript1 />
    }[language.toLowerCase()]

    return cloneElement(selectedLanguage || <MdCode />, {
      size: 16,
      className: 'text-shape_color_lightTheme dark:text-shape_color_darkTheme'
    })
  }

  useEffect(() => {
    const timer = setTimeout(() => {
      setActiveItem((prevState) => {
        if (prevState + 1 === items.length) return 0
        else return prevState + 1
      })
    }, 5000)

    return () => clearTimeout(timer)
  }, [activeItem])

  return (
    <div className="grid grid-cols-[auto,_1fr] gap-6 relative h-full w-full overflow-hidden pb-5 before:content-[''] before:absolute before:bottom-0 before:left-0 before:w-full before:h-5 before:z-10 before:bg-gradient-to-t from-background_color_lightTheme to-transparent dark:from-background_color_darkTheme">
      <div className="flex flex-col justify-center gap-6 z-10">
        {items.map((item, index) => (
          <motion.button
            key={item.id}
            initial={verticalHidden}
            animate={verticalVisible}
            transition={{ delay: index * 0.1 }}
            className={`w-5 h-5 rounded-[50%] transition-colors ${
              index === activeItem
                ? `bg-${color}-600 pointer-events-none`
                : `bg-contrast_color_lightTheme dark:bg-contrast_color_darkTheme hover:bg-${color}-600`
            }`}
            onClick={() => setActiveItem(index)}
          />
        ))}
      </div>

      <div className="flex items-center justify-center relative h-96 lg:h-full w-full">
        <AnimatePresence mode="wait">
          {items.map(
            (item, index) =>
              index === activeItem && (
                <motion.div
                  key={item.id}
                  initial={verticalHidden}
                  animate={verticalVisible}
                  exit={verticalHidden}
                  className={`flex flex-col justify-between items-start absolute text-left h-full w-full p-12 rounded-3xl bg-shape_color_lightTheme dark:bg-shape_color_darkTheme z-0 ${
                    index !== activeItem && 'top-[100%]'
                  }`}
                >
                  <div className="self-end flex items-center gap-2">
                    <Image
                      src={item.owner.avatar_url}
                      width={500}
                      height={500}
                      alt={`Avatar de ${item.owner.login}`}
                      className="w-12 rounded-[50%]"
                    />

                    <span className="font-bold text-contrast_color_lightTheme dark:text-contrast_color_darkTheme">
                      {item.owner.login}
                    </span>
                  </div>

                  <div className="flex flex-col gap-4">
                    <div className="flex flex-col gap-2">
                      <h2 className="text-4xl font-bold text-contrast_color_lightTheme dark:text-contrast_color_darkTheme">
                        {item.name}
                      </h2>

                      <div className="flex flex-wrap items-center gap-2">
                        <div className="flex items-center gap-1 px-1.5 rounded-[0.375rem] bg-contrast_color_lightTheme dark:bg-contrast_color_darkTheme">
                          <MdStar
                            size={16}
                            className="text-shape_color_lightTheme dark:text-shape_color_darkTheme"
                          />

                          <span className="text-shape_color_lightTheme dark:text-shape_color_darkTheme text-base">
                            {item.stargazers_count}
                          </span>
                        </div>

                        {item.language && (
                          <div className="flex items-center gap-1 px-1.5 rounded-[0.375rem] bg-contrast_color_lightTheme dark:bg-contrast_color_darkTheme">
                            {handleLanguageIcon(item.language)}

                            <span className="text-shape_color_lightTheme dark:text-shape_color_darkTheme text-base">
                              {item.language}
                            </span>
                          </div>
                        )}
                      </div>

                      <div className="flex items-center gap-2">
                        <a
                          href={item.html_url}
                          target="_blank"
                          className={`text-${color}-600 font-bold transition-colors hover:text-${color}-900`}
                        >
                          Repository
                        </a>

                        {item.homepage && (
                          <>
                            <span className="w-2 h-2 rounded-[50%] bg-contrast_color_lightTheme dark:bg-contrast_color_darkTheme" />

                            <a
                              href={item.homepage}
                              target="_blank"
                              className={`text-${color}-600 font-bold transition-colors hover:text-${color}-900`}
                            >
                              Homepage
                            </a>
                          </>
                        )}
                      </div>
                    </div>

                    <p className="text-contrast_color_lightTheme dark:text-contrast_color_darkTheme">
                      {item.description}
                    </p>
                  </div>
                </motion.div>
              )
          )}
        </AnimatePresence>
      </div>
    </div>
  )
}
