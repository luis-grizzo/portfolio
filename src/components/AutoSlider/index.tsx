'use client'

import { useEffect, useRef, useState } from 'react'
import { AnimatePresence, motion } from 'framer-motion'

import { timeUnits } from '@/constants/timeUnits'

import { defaultTransition } from '@/utils/animations'

interface AutoSliderProps {
  children: React.ReactNode
}

type FullScrollTypes = 'left' | 'right' | null

export function AutoSlider({ children }: AutoSliderProps) {
  const containerRef = useRef<HTMLDivElement>(null)
  const contentRef = useRef<HTMLDivElement>(null)

  const [fullScrolled, setFullScrolled] = useState<FullScrollTypes>('left')

  function handleScroll() {
    const { current: container } = containerRef
    const { current: content } = contentRef

    if (container && content) {
      let value!: FullScrollTypes

      if (container.scrollLeft === 0) value = 'left'
      else if (
        container.scrollLeft ===
        content.clientWidth - container.clientWidth
      )
        value = 'right'
      else value = null

      setFullScrolled(value)
    }
  }

  function handleAnimation() {
    const { current: container } = containerRef
    const { current: content } = contentRef

    if (container && content) {
      let destination!: number

      if (container.scrollLeft === 0)
        destination = content.clientWidth - container.clientWidth
      else destination = 0

      container.scrollTo({
        left: destination,
        behavior: 'smooth'
      })
    }
  }

  useEffect(() => {
    const { current: container } = containerRef

    if (container) container.addEventListener('scroll', handleScroll)

    const timer = setInterval(handleAnimation, timeUnits.second * 3)

    return () => {
      if (container) container.removeEventListener('scroll', handleScroll)

      clearInterval(timer)
    }
  }, [])

  return (
    <div className="relative w-full">
      <AnimatePresence mode="popLayout">
        {fullScrolled !== 'left' && (
          <motion.div
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            transition={defaultTransition}
            key="left"
            data-testid="left-fade"
            className="absolute left-0 top-0 w-4 h-full bg-gradient-to-r from-neutral-950 z-50"
          />
        )}

        {fullScrolled !== 'right' && (
          <motion.div
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            transition={defaultTransition}
            key="right"
            data-testid="right-fade"
            className="absolute right-0 top-0 w-4 h-full bg-gradient-to-l from-neutral-950 z-50"
          />
        )}
      </AnimatePresence>

      <div ref={containerRef} role="tabpanel" className="overflow-hidden -z-50">
        <div
          ref={contentRef}
          role="tablist"
          className="flex items-center gap-4 w-min"
        >
          {children}
        </div>
      </div>
    </div>
  )
}
