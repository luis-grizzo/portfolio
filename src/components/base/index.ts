export * from './Card'
export * from './LinkButton'
export * from './Logo'
