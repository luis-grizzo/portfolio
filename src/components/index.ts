export * from './AutoSlider'
export * from './Button'
export * from './Card'
export * from './Footer'
export * from './Footnote'
export * from './Hero'
export * from './LinkButton'
export * from './LittleSunshine'
export * from './Navbar'
