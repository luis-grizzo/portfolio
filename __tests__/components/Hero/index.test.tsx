import { render, screen } from '@testing-library/react'

import { Hero } from '@/components'

import type { Random } from 'unsplash-js/dist/methods/photos/types'

describe('Hero', () => {
  const mockImage: Random = {
    alt_description: 'alt_description',
    blur_hash: null,
    color: null,
    created_at: '2024-05-21T14:42:05Z',
    description: null,
    exif: {
      aperture: null,
      exposure_time: null,
      focal_length: null,
      iso: null,
      make: null,
      model: null
    },
    height: 100,
    id: 'id',
    likes: 0,
    links: {
      download: 'download',
      download_location: 'download_location',
      html: 'html',
      self: 'self'
    },
    location: {
      city: null,
      country: null,
      name: null,
      position: { latitude: null, longitude: null }
    },
    promoted_at: null,
    updated_at: '2024-07-03T18:25:47Z',
    urls: {
      full: '/full',
      raw: '/raw',
      regular: '/regular',
      small: '/small',
      thumb: '/thumb'
    },
    user: {
      bio: null,
      first_name: 'first_name',
      id: 'id',
      instagram_username: null,
      last_name: null,
      links: {
        followers: 'followers',
        following: 'following',
        html: 'html',
        likes: 'likes',
        photos: 'photos',
        portfolio: 'portfolio',
        self: 'self'
      },
      location: null,
      name: 'name',
      portfolio_url: null,
      profile_image: { large: 'large', medium: 'medium', small: 'small' },
      total_collections: 0,
      total_likes: 0,
      total_photos: 0,
      twitter_username: null,
      updated_at: '2024-07-03T18:25:47Z',
      username: 'username'
    },
    width: 100
  }

  it('should match snapshot', () => {
    const { container } = render(<Hero image={mockImage} />)

    expect(container).toMatchSnapshot()
  })

  it('should render a generic alt description when the info is not available', () => {
    const newMockImage: Random = { ...mockImage, alt_description: null }

    render(<Hero image={newMockImage} />)

    const eImage = screen.getByRole('img') as HTMLImageElement

    expect(eImage.alt).toEqual('Generic background image')
  })
})
