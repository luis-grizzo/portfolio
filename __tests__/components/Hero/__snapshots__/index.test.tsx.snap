// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero should match snapshot 1`] = `
<div>
  <header
    class="relative flex flex-col items-center justify-center w-full h-[75vh] md:h-[70vh] lg:h-[65vh] border-b-1 border-neutral-400/10"
  >
    <div
      class="container mx-auto px-4 py-8"
    >
      <div
        class="flex flex-col justify-center gap-4 md:max-w-2xl lg:max-w-3xl xl:max-w-4xl"
      >
        <h1
          class="text-4xl sm:text-5xl text-pretty"
          style="opacity: 0; transform: translateX(-50px);"
        >
          Hi, i'm a Front-end Developer.
        </h1>
        <p
          class="text-xl sm:text-2xl text-pretty text-neutral-300"
          style="opacity: 0; transform: translateX(-50px);"
        >
          Welcome to my portfolio, explore my projects and get in touch!
        </p>
      </div>
    </div>
    <span
      class="hidden md:block absolute top-8 left-1/2 -translate-x-1/2 text-base py-4 text-neutral-300"
    >
      luisgrizzo.dev
    </span>
    <figure
      class="absolute top-0 left-0 w-full h-full opacity-30 overflow-hidden -z-50"
    >
      <img
        alt="alt_description"
        class="w-full h-[calc(100%_+_300px)] object-cover object-center"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="100"
        src="/_next/image?url=%2Ffull&w=256&q=75"
        srcset="/_next/image?url=%2Ffull&w=128&q=75 1x, /_next/image?url=%2Ffull&w=256&q=75 2x"
        style="color: transparent; opacity: 0; will-change: opacity; transform: translateY(-300px) scaleX(1.2) scaleY(1.2);"
        width="100"
      />
    </figure>
    <span
      class="absolute bottom-8 px-4 text-xs text-center text-pretty text-neutral-300"
    >
      Photo by 
      <a
        class="underline underline-offset-2 text-blue-400 hover:text-blue-200 transition-colors"
        href="https://unsplash.com/@username?utm_source=Unconscious&utm_medium=referral"
        rel="noreferrer"
        target="_blank"
      >
        name
      </a>
       on 
      <a
        class="underline underline-offset-2 text-blue-400 hover:text-blue-200 transition-colors"
        href="https://unsplash.com/?utm_source=Unconscious&utm_medium=referral"
        rel="noreferrer"
        target="_blank"
      >
        Unsplash
      </a>
    </span>
  </header>
</div>
`;
